test/Test/Database/LSMTree/Internal/Run.hs

```diff
@@ -0,0 +1,163 @@
+      ]
+    ]
+  where
+    withSessionDir = withTempDirectory "" "session"
+
+-- | Runs in IO, with a real file system.
+testSingleInsert :: FilePath -> Key -> Val -> Maybe Blob -> IO ()
+testSingleInsert sessionRoot key val mblob = do
+    let fs = FsIO.ioHasFS (FS.MountPoint sessionRoot)
+    -- flush write buffer
+    let wb = WB.addEntryNormal key (N.Insert val mblob) WB.empty
+    _ <- fromWriteBuffer fs (RunFsPaths 42) wb
+    -- check all files have been written
+    bsFilter <- BS.readFile (sessionRoot </> "active" </> "42.filter")
+    mempty @=? bsFilter  -- TODO: empty for now, should be written later
+    bsIndex <- BS.readFile (sessionRoot </> "active" </> "42.index")
+    mempty @=? bsIndex   -- TODO: empty for now, should be written later
+    bsBlobs <- BS.readFile (sessionRoot </> "active" </> "42.blobs")
+    bsKops <- BS.readFile (sessionRoot </> "active" </> "42.keyops")
+    let page = rawPageFromByteString bsKops 0
+    -- check page
+    1 @=? rawPageNumKeys page
+    let SerialisedKey' key' = serialiseKey key
+    let SerialisedValue' val' = serialiseValue val
+    case mblob of
+      Nothing -> do
+        0 @=? rawPageNumBlobs page
+        mempty @=? bsBlobs
+        Just (Insert val') @=? rawPageEntry page key'
+      Just b -> do
+        1 @=? rawPageNumBlobs page
+        let entry = fmap (readBlob bsBlobs) <$> rawPageEntry page key'
+        Just (InsertWithBlob val' (serialiseBlob b)) @=? entry
+
+-- | Runs in IO, but using a mock file system.
+--
+-- TODO: Also test file system errors.
+prop_WriteAndRead :: WriteBuffer Key Val Blob -> Property
+prop_WriteAndRead wb = ioProperty $ do
+    fs <- FsSim.mkSimErrorHasFS' FsSim.empty FsSim.emptyErrors
+    -- flush write buffer
+    let fsPaths = RunFsPaths 42
+    _ <- fromWriteBuffer fs fsPaths wb
+    -- read pages
+    bsBlobs <- getFile fs (runBlobPath fsPaths)
+    bsKops <- getFile fs (runKOpsPath fsPaths)
+    let pages = rawPageFromByteString bsKops <$> [0, 4096 .. (BS.length bsKops - 1)]
+    -- check pages
+    return $ label ("Number of pages: " <> showPowersOf10 (length pages)) $ do
+      let vals = concatMap (bifoldMap pure mempty . snd) (WB.content wb)
+      tabulate "Value size" (map (showPowersOf10 . sizeofValue) vals) $
+        pagesContainEntries bsBlobs pages (WB.content wb)
+  where
+    getFile fs path = FS.withFile fs path FS.ReadMode $ \h -> do
+      n <- FS.hGetSize fs h
+      getFile' fs h n
+    getFile' fs h n
+      | n <= 0 = return ""
+      | otherwise = do
+          bs <- FS.hGetSome fs h n
+          (bs <>) <$> getFile' fs h (n - fromIntegral (BS.length bs))
+
+pagesContainEntries :: ByteString -> [RawPage] ->
+                       [(SerialisedKey, Entry SerialisedValue SerialisedBlob)] ->
+                       Property
+pagesContainEntries _ [] es = counterexample ("k/ops left: " <> show es) (null es)
+pagesContainEntries bsBlobs (page : pages) kops = do
```

Comment:
This looks similar to the "QC test for run construction and lookup" that I've been working on. Once both are in place, we can see whether we can unify some of it